src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/ReadsPipelineSpark.java

```diff
@@ -0,0 +1,118 @@
+        usageExample = "Hellbender ReadsPreprocessingPipeline -I single.bam -R referenceURL -BQSRKnownVariants variants.vcf -O output.bam",
+        programGroup = DataFlowProgramGroup.class
+)
+
+/**
+ * ReadsPreprocessingPipeline is our standard pipeline that takes aligned reads (likely from BWA) and runs MarkDuplicates
+ * and BQSR. The final result is analysis-ready reads.
+ */
+public class ReadsPipelineSpark extends SparkCommandLineProgram {
+    private static final long serialVersionUID = 1L;
+
+    @Argument(doc = "uri for the input bam, either a local file path or a gs:// bucket path",
+            shortName = StandardArgumentDefinitions.INPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.INPUT_LONG_NAME,
+            optional = false)
+    protected String bam;
+
+    @Argument(doc = "the output bam", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
+    protected String output;
+
+    @Argument(doc = "the reference name", shortName = StandardArgumentDefinitions.REFERENCE_SHORT_NAME,
+            fullName = StandardArgumentDefinitions.REFERENCE_LONG_NAME, optional = false)
+    protected String referenceURL;
+
+    @Argument(doc = "the known variants", shortName = "BQSRKnownVariants", fullName = "baseRecalibrationKnownVariants", optional = true)
+    protected List<String> baseRecalibrationKnownVariants;
+
+    @ArgumentCollection
+    protected IntervalArgumentCollection intervalArgumentCollection = new OptionalIntervalArgumentCollection();
+
+    @Override
+    protected void runPipeline(final JavaSparkContext ctx) {
+        ReadsSparkSource readSource = new ReadsSparkSource(ctx);
+        SAMFileHeader readsHeader = readSource.getHeader(bam);
+        final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(readsHeader.getSequenceDictionary())
+                : IntervalUtils.getAllIntervalsForReference(readsHeader.getSequenceDictionary());
+        JavaRDD<GATKRead> initialReads = readSource.getParallelReads(bam, intervals);
+
+        JavaRDD<GATKRead> markedReads = initialReads.map(new MarkDuplicatesStub());
+        VariantsSparkSource variantsSparkSource = new VariantsSparkSource(ctx);
+        JavaRDD<Variant> variants = variantsSparkSource.getParallelVariants(baseRecalibrationKnownVariants);
+
+        GCSOptions options = PipelineOptionsFactory.as(GCSOptions.class);
+        options.setApiKey(apiKey);
+
+        final ReferenceDataflowSource referenceDataflowSource = new ReferenceDataflowSource(options, referenceURL, BaseRecalibratorDataflow.BQSR_REFERENCE_WINDOW_FUNCTION);
+
+        JavaPairRDD<GATKRead, ReadContextData> rddReadContext = AddContextDataToReadSpark.add(markedReads, referenceDataflowSource, variants);
+
+        JavaRDD<GATKRead> finalReads = rddReadContext.map(new ApplyBQSRStub());
```

Comment:
Agree this is worth exploring, but recommend doing it as an isolated performance experiment after the full version of the existing spark pipeline is merged (so we can compare with the existing approach in `AddContextDataToReadSpark`)